spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/DVIterable.java

```diff
@@ -0,0 +1,147 @@
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.MetadataColumns;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.deletes.PositionDeleteIndex;
+import org.apache.iceberg.io.CloseableGroup;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.puffin.BlobMetadata;
+import org.apache.iceberg.puffin.Puffin;
+import org.apache.iceberg.puffin.PuffinReader;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Pair;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.unsafe.types.UTF8String;
+
+class DVIterable extends CloseableGroup implements CloseableIterable<InternalRow> {
+  private final Puffin.ReadBuilder builder;
+  private final PartitionSpec spec;
+  private final DeleteFile deleteFile;
+  private final Schema projection;
+
+  DVIterable(InputFile inputFile, DeleteFile deleteFile, PartitionSpec spec, Schema projection) {
+    this.deleteFile = deleteFile;
+    this.builder = Puffin.read(inputFile);
+    this.spec = spec;
+    this.projection = projection;
+  }
+
+  @Override
+  public CloseableIterator<InternalRow> iterator() {
+    PuffinReader reader = builder.build();
+    addCloseable(reader);
+    return new DVIterator(reader);
+  }
+
+  private class DVIterator implements CloseableIterator<InternalRow> {
+    private final PuffinReader reader;
+    private final Iterator<Pair<BlobMetadata, ByteBuffer>> blobs;
+    private Pair<BlobMetadata, ByteBuffer> currentBlob;
+    private Iterator<Long> positions;
+
+    DVIterator(PuffinReader reader) {
+      this.reader = reader;
+      try {
+        blobs = reader.readAll(reader.fileMetadata().blobs()).iterator();
+        advanceBlob();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private void advanceBlob() {
+      if (blobs.hasNext()) {
+        this.currentBlob = blobs.next();
+        List<Long> pos = Lists.newArrayList();
+        PositionDeleteIndex.deserialize(currentBlob.second().array(), deleteFile).forEach(pos::add);
+        this.positions = pos.iterator();
+      } else {
+        this.currentBlob = null;
+      }
+    }
+
+    @Override
+    public boolean hasNext() {
+      boolean hasNextPosition = positions.hasNext();
+      if (!hasNextPosition) {
+        advanceBlob();
+      }
+
+      return null != currentBlob && hasNextPosition;
+    }
+
+    @Override
+    public InternalRow next() {
+      long position = positions.next();
+
+      List<Object> values = Lists.newArrayList();
```

Comment:
This seems fairly expensive to be called for each record, especially given that all columns are constants, except the position.

Reply:
good point, I've updated this so that we reuse the current row and always only update the position